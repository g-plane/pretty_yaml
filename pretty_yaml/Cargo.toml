[package]
name = "pretty_yaml"
version = "0.0.0"
edition = "2021"
authors = ["Pig Fang <g-plane@hotmail.com>"]
description = "Configurable YAML formatter."
repository = "https://github.com/g-plane/pretty_yaml"
license = "MIT"
exclude = ["/tests"]

[dependencies]
itertools = "0.12"
serde = { version = "1.0", optional = true }
tiny_pretty = "0.1"
yaml_parser = { version = "0.0", path = "../yaml_parser" }

[dev-dependencies]
insta = { version = "1.38", features = ["glob"] }
similar-asserts = "1.5"
toml = "0.8"

[features]
config_serde = ["serde"]
