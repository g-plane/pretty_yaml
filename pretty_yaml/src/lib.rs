#![doc = include_str!("../README.md")]

use crate::{
    config::FormatOptions,
    printer::{Ctx, DocGen},
};
use tiny_pretty::{print, IndentKind, PrintOptions};
use yaml_parser::{
    ast::{AstNode, Root},
    SyntaxError,
};

pub mod config;
mod printer;

/// Format the given source input.
pub fn format_text(input: &str, options: &FormatOptions) -> Result<String, SyntaxError> {
    let syntax = yaml_parser::parse(input)?;
    let root = Root::cast(syntax).expect("expected root node");
    Ok(print_tree(&root, options))
}

/// Print the given concrete syntax tree.
/// You may use this when you already have the parsed CST.
pub fn print_tree(root: &Root, options: &FormatOptions) -> String {
    let ctx = Ctx {
        indent_width: options.layout.indent_width,
        options: &options.language,
    };
    print(
        &root.doc(&ctx),
        &PrintOptions {
            indent_kind: IndentKind::Space,
            line_break: options.layout.line_break.clone().into(),
            width: options.layout.print_width,
            tab_size: options.layout.indent_width,
        },
    )
}
