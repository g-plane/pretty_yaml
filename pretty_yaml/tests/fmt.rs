use insta::{assert_snapshot, glob, Settings};
use pretty_yaml::{config::FormatOptions, format_text};
use std::{collections::HashMap, fs, path::Path};

#[test]
fn fmt_snapshot() {
    glob!("fmt/**/*.yaml", |path| {
        let input = fs::read_to_string(path).unwrap();

        let options = fs::read_to_string(path.with_file_name("config.toml"))
            .map(|config_file| {
                toml::from_str::<HashMap<String, FormatOptions>>(&config_file).unwrap()
            })
            .ok();

        if let Some(options) = options {
            options.into_iter().for_each(|(option_name, options)| {
                let output = run_format_test(path, &input, &options);
                build_settings(path).bind(|| {
                    let name = path.file_stem().unwrap().to_str().unwrap();
                    assert_snapshot!(format!("{name}.{option_name}"), output);
                });
            })
        } else {
            let output = run_format_test(path, &input, &Default::default());
            build_settings(path).bind(|| {
                let name = path.file_stem().unwrap().to_str().unwrap();
                assert_snapshot!(name, output);
            });
        }
    });
}

fn run_format_test(path: &Path, input: &str, options: &FormatOptions) -> String {
    let output = format_text(&input, &options)
        .map_err(|err| format!("failed to format '{}': {:?}", path.display(), err))
        .unwrap();
    assert!(
        !output.contains(" \n"),
        "'{}' has trailing whitespace",
        path.display()
    );
    let regression_format = format_text(&output, &options)
        .map_err(|err| {
            format!(
                "syntax error in stability test '{}': {:?}",
                path.display(),
                err
            )
        })
        .unwrap();
    similar_asserts::assert_eq!(
        output,
        regression_format,
        "'{}' format is unstable",
        path.display()
    );

    output
}

fn build_settings(path: &Path) -> Settings {
    let mut settings = Settings::clone_current();
    settings.set_snapshot_path(path.parent().unwrap());
    settings.remove_snapshot_suffix();
    settings.set_prepend_module_to_snapshot(false);
    settings.remove_input_file();
    settings.set_omit_expression(true);
    settings.remove_input_file();
    settings.remove_info();
    settings
}
