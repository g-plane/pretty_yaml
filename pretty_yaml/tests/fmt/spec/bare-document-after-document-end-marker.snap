---
source: pretty_yaml/tests/fmt.rs
---
---
scalar1
...
key: value
